packages/evm/src/eof/container.ts

```diff
@@ -0,0 +1,461 @@
+   * Verify that the current uint8 pointed to by the pointer is the expected uint8
+   * Also updates the pointer
+   * @param expect The uint to expect
+   * @param errorStr Optional error string when the read uint is not the expected uint
+   */
+  verifyUint(expect: number, errorStr?: string) {
+    if (this.readUint() !== expect) {
+      validationError(EOFError.VerifyUint, this.ptr - 1, errorStr)
+    }
+  }
+
+  /**
+   * Same as readUint, except this reads an uint16
+   * @param errorStr
+   * @returns
+   */
+  readUint16(errorStr?: string) {
+    const end = this.ptr + 2
+    if (end > this.data.length) {
+      validationError(EOFError.OutOfBounds, this.ptr, errorStr)
+    }
+    const ptr = this.ptr
+    this.ptr += 2
+    return new DataView(this.data.buffer).getUint16(ptr)
+  }
+
+  /**
+   * Get the current pointer of the stream
+   * @returns The pointer
+   */
+  getPtr() {
+    return this.ptr
+  }
+
+  // Get the remainder bytes of the current stream
+  readRemainder() {
+    return this.data.slice(this.ptr)
+  }
+
+  // Returns `true` if the stream is fully read, or false if there are dangling bytes
+  isAtEnd() {
+    return this.ptr === this.data.length
+  }
+}
+
+// TODO add initcode flags (isEOFContract)
+// TODO validation: mark sections as either initcode or runtime code to validate
+
+/**
+ * The EOFHeader, describing the header of the EOF container
+ */
+class EOFHeader {
+  typeSize: number // Size of the types section
+  codeSizes: number[] // Sizes of the code sections
+  containerSizes: number[] // Sizes of the containers
+  dataSize: number // Size of the data section
+  dataSizePtr: number // Used to edit the dataSize in RETURNCONTRACT
+  buffer: Uint8Array // The raw buffer of the entire header
+
+  private codeStartPos: number[] // Internal array to track at which byte of the container the code starts (per section)
+
+  /**
+   * Create an EOF header. Performs various validation checks inside the constructor
+   * @param input The input should either be a raw header, or a complete container
+   */
+  constructor(input: Uint8Array) {
+    if (input.length > MAX_HEADER_SIZE) {
+      throw new Error('err: container size more than maximum valid size')
+    }
+    const stream = new StreamReader(input)
+    // Verify that the header starts with 0xEF0001
+    stream.verifyUint(FORMAT, EOFError.FORMAT)
+    stream.verifyUint(MAGIC, EOFError.MAGIC)
+    stream.verifyUint(VERSION, EOFError.VERSION)
+    if (input.length < 15) {
+      throw new Error('err: container size less than minimum valid size')
+    }
+    // Verify that the types section is present, and verify that the type section length is valid
+    stream.verifyUint(KIND_TYPE, EOFError.KIND_TYPE)
+    const typeSize = stream.readUint16(EOFError.TypeSize)
+    if (typeSize < TYPE_MIN) {
+      validationError(EOFError.InvalidTypeSize, typeSize)
+    }
+    if (typeSize % TYPE_DIVISOR !== 0) {
+      validationError(EOFError.InvalidTypeSize, typeSize)
+    }
```

Comment:
Total nit but can we combine these since the validation error is the same?
```suggestion
    if ((typeSize < TYPE_MIN) || (typeSize % TYPE_DIVISOR !== 0))  {
      validationError(EOFError.InvalidTypeSize, typeSize)
    }
```